### ISSUE TITLE
Styled netG, Unstyled Images?

### ISSUE THREAD

**Issue by denw** [sidward14/Style-AttnGAN#6]
Hi Sidhartha --

We're doing a final project for school class (NCSU).  We used your adaptation of the model to make images from the oxford 102 flowers dataset.  It's great thanks a lot!

We've trained the model using the style netG.

**Is there any way now to get un-styled images without retraining**?  i.e. we have a netG.pth that evaluates un-styled for 64 and 128 images and styled images for the 256 images.  I'd like to also generate the unstyled 256 images  Is there a way to do this without re-training w/ GAN.B_STYLEGEN = False?

(we all have weak computers and we couldn't get it to run on the school servers, so we had to rent a server to run on which is why we don't want to re-train)

(I /think/ I saw unstyled 256 images in the training output images attention maps.  I /think/ the images are styled ha - they sure look different from the 64/128 imgs.  Maybe this question is way out in left field.)

Thanks!
Dennis

**Comment by sidward14** [comment]
Hello! I'm glad my repo is helping you all out with your final project!

So the styling is basically everything that contributes to variation in the image generation (there's some stochastic noise, but that's only for very fine-grained details). These "styles" are a result of the sentence embedding from the text encoder and a latent code sampled from a Gaussian distribution. So without the styling, you won't be taking the input captions into account nor will you be taking into account the latent code. All the generated images would look more-or-less the same. Is this what you're looking for, or something else?

**Comment by denw** [comment]
In the training output images, some of the images looked amazingly realistic, with really fine detail with veins in the petals for example.  Then the final styled images looked more like they had a "style" applied.  Like the edges of the petals had a bright ring around them and the detail was less defined.  The top images do have the input from the captions - like they will be "spiky" petals for example (although "spiky" doesn't look great, but it clearly makes a difference in the detail that is added)

Examples in the training output images:
https://drive.google.com/file/d/1GsdXIigBtxov9u5kj9sinQpNxxhKEBcl/view?usp=sharing
https://drive.google.com/file/d/1W4XxVl9cVGvC95UiY3Y738g9cGdWUzFE/view?usp=sharing

I would love to see a 256 version of the top images in each row, if it is possible (without retraining).  From your answer, it sounds like not.  (it was a long shot)   :-)



As an aside, when setting up the training, we accidentally set one of the parameters to 64 while debugging and left it (was defaulted to I think 150 and had been 299 - maybe the latent Z dim?  - I don't have in front of me right now).  Didn't notice it until halfway through training.  I wonder if that gave us less variation in the flowers that came out and slower learning.  But it does do color, it will try to make spots (they come out overly regular, like rows and columns of dots but it is doing something), and it will try to make "spiky", "hairy", or "thin" petals when those words appear in the caption (although they also come out vertical or horizontal, but not radial).  But the overall shape is always a central circular section with radial petals emanating.  A top-view.  Doesn't output "trumpet" shape or other shapes.  I think this is because the text captions don't provide consistently used or unique terms for the different shapes.

If there are two flowers in 64 image, it morphs the petals in between them into a single petal which is not expected but kind of cool.


If you like we can submit a pull request with our configurations for the flowers data that we ended up using.

**Comment by sidward14** [comment]
Ahh I see the problem now. Well first off, very happy it works well for flowers. Yes, you will get this sort of painted look sometimes for the higher resolutions in style-based GANs. In my experience, simply training longer fixes this issue (256 is 4x the number of pixels after all). You have the ability to continue training from where you left off. Unfortunately, there's no quick and easy fix to this, you will have to train some more.

As for the parameter you guys set to 64, was that for pretraining the DAMSM or for the training the GAN? If the former, keep it at 299, because the [CNN_ENCODER](https://github.com/sidward14/Style-AttnGAN/blob/8aa8da6601725c425e272f0bededa32033a38d84/code/model.py#L223) used by the DAMSM is set to resize your image to 299 res (and this particular encoder, inception v3, also happens to work best for 299 res). So yes, that's probably why the learning for you was slower. You can submit a PR and I can take a look if you want.

Again, very happy to see it working well for flowers. Actually, I've been so busy, but I want to modernize some of the models in here (image encoder, StylGAN3, etc), and then maybe even submit a paper...

**Comment by denw** [comment]
Okay got it.

And after I reviewed our configuration I realized it was TREE.BASE_SIZE for the DAMSM encoder.  I started with 64 and it did run the GAN training -- I realized re-pre-trained w/ 299, then restarted the GAN training with the 299 DAMSM model.  It thankfully picked right up where it was.  Maybe that also gave our training a slow start.

Also when we looked more closely at the training output images, some of them were nicer with the top image.  So you are right we probably just needed to train more.

I'll mark this issue solved thanks so much for taking a look at our issue.

If you are open to it, we will submit some requests with additions of our .cfg files where people can try this awesome architecture with the flowers database as well.

Our project angle was we would integrate word2vec encodings into the text encoding but our project just didn't have enough time.  We barely got it to train lol.  So we ended up just using word2vec to replace words in a example captions that were not in the training data set with the nearest word found by our wor2vec model that were in the data set.  Even though we only have a surface understanding of the network we still learned a lot.

Thanks,
Dennis, Jacob, Benjamin

**Comment by sidward14** [comment]
That incredible! Your word2vec work is very interesting! I wonder how it went...the embeddings learned per token in the text encoder are a bit different than those you would get from word2vec, because from backprop, the text encoder "knows" that it's trying to match with image pixels (which is not the case in word2vec). Also the idea of integrating word2vec to strengthen the text encoder sounds like a good idea.

And yes, please do submit your PRs  with the flowers dataset! I would love to integrate in the config files that worked best for you and show off some of your results in the README!
